### ISSUE TITLE
Why does "1176413S03" get converted to numeric when using `readr::type_convert` ?

### ISSUE THREAD

**Issue by shahronak47** [tidyverse/readr#1562]
Why does these values get converted into numeric when using `readr::type_convert` ? I would expect them to stay characters. 

```
x <- c("1176413S03", "1176413S06", "1176413S02", "1176413S08", "1176413S05", "1176413S04")
df <- data.frame(x)
str(df)

'data.frame':	6 obs. of  1 variable:
 $ x: chr  "1176413S03" "1176413S06" "1176413S02" "1176413S08" ...

df1 <- readr::type_convert(df)
str(df1)

'data.frame':	6 obs. of  1 variable:
 $ x: num  1.18e+09 1.18e+12 1.18e+08 1.18e+14 1.18e+11 ...
```

**Comment by joranE** [comment]
This behavior appears to happen only for the letters D, E, F, L & S (upper or lower case). E makes sense.

```
res <- purrr::map_chr(
  .x = LETTERS,
  .f = \(x) {
    d <- readr::type_convert(
      df = data.frame(
        x = glue::glue("1176413{letter}03",letter = x)
      )
    )
    class(d$x)
  }
)

setNames(res,LETTERS)

> setNames(res,LETTERS)
          A           B           C           D           E           F 
"character" "character" "character"   "numeric"   "numeric"   "numeric" 
          G           H           I           J           K           L 
"character" "character" "character" "character" "character"   "numeric" 
          M           N           O           P           Q           R 
"character" "character" "character" "character" "character" "character" 
          S           T           U           V           W           X 
  "numeric" "character" "character" "character" "character" "character" 
          Y           Z 
"character" "character" 
```

**Comment by chainsawriot** [comment]
https://github.com/tidyverse/readr/blob/96ddac314b47402bc63e1f81c149c463cf58e3da/src/QiParsers.h#L157-L181

And `vroom` doesn't use qiparser

https://github.com/tidyverse/vroom/blob/73c90c4fe490c0588b20ac527c40fcb1c683683e/src/vroom_num.cc#L62

**Comment by jxu** [comment]
I think these may be for C floating point constants https://en.cppreference.com/w/c/language/floating_constant

e is exponent for decimal floating point
p is exponent for hex floating point (rare)
f is suffix for float
l is suffix for long double (rare)

I don't know what s is, and a parser should really only be using e or no letters.

**Comment by joranE** [comment]
At the _very_ least it should be documented, both in `?type_convert` and there should be at least some reference to this behavior in `?read_csv` & friends as well.

**Comment by jxu** [comment]
I agree, but looking at the state of the issue tracker and other reports listed I don't think it will be addressed anytime soon

**Comment by tjmahr** [comment]
If F/L/S/f/l/s are suffixes, per https://en.cppreference.com/w/c/language/floating_constant, then they are not numbers if they have digits after them, and the `bsd_strtod` implentation is wrong. 

`bsd_strtod()` was taken from FreeBSD and that port was based on the 386BSD implementation. My hunch is that F/L/S/f/l/s were also [meant to be suffixes](https://github.com/386bsd/386bsd/blob/9eaf3b3359255f461055b8f582131dd4695e750d/usr/src/usr.bin/gas/config/atof-ieee.c#L174-L204) there as well, and `eE` [were intended](https://github.com/386bsd/386bsd/blob/9eaf3b3359255f461055b8f582131dd4695e750d/usr/src/usr.bin/gas/config/i386.c#L66) to be the exponent markers. But something happened during the porting, or people actually wrote numbers with F/L/S as exponent markers and the port author wanted to cover them.

Still, we can just fix this and use only e/E/p/P as exponent markers.

**Comment by bbolker** [comment]
IMO adding p/P as exponent markers is **not** a good idea; according to the spec, p/P are supposed to be used in a hexadecimal context to signify powers of 2 (i.e. `0xap2` = `hex(a)*2^hex(2)` = 40 (decimal). It's not obvious whether it even makes sense in a non-hexadecimal context (is `55p2` equal to `55*2^2` ?), and in any case the machinery here isn't handling hexadecimal values as far as I know.

**Comment by tjmahr** [comment]
Yes, that's true.

```
> 0xa
[1] 10
> 0xap1
[1] 20
> 0xap2
[1] 40

> readr::parse_number("0xa")
[1] 0
> readr::parse_number("0xap2")
[1] 0
```

**Comment by jennybc** [comment]
Also beware that `parse_*()` functions use the 1st edition parser, which is not the default for parsing a delimited file 🤪. See #1617.

**Comment by bbolker** [comment]
Following up @jennybc's comment: `type_convert` and `parse_number` don't necessarily give the same results.

```r
library(readr)
tc <- function(x) suppressMessages(data.frame(x=x) |> type_convert() |> dplyr::pull())
tc("3s03")
## [1] 3000
parse_number("3s03")
## [1] 3
tc("0xa3")
## [1] "0xa3"
parse_number("0xa3")
## [1] 0
```

**Comment by jennybc** [comment]
Re: `type_convert()` vs. `parse_number()`, this is really apples to oranges. This isn't directly related the main topic here -- they use completely different parsers. When `type_convert()` guesses a column is numeric, it parses with `parse_double()`, which uses the Qi parser and understands these exponent/suffix letters (for better or for worse). `parse_number()` is a different, more special-purpose function designed to extract numbers from messy strings (like "1,234.56").

Exploring these examples a bit more and making sure to include `readr::read_*()`, which is just a call to `vroom::vroom()` by default. `type_convert()` and the `parse_*()` functions really are sort of an evolutionary dead end, in the sense that they exercise 1st edition code at this point.

``` r
library(readr)

# guess type, then parse
tc <- function(x) {
  suppressMessages(data.frame(x = x) |> type_convert() |> dplyr::pull())
}
vr <- function(x) {
  read_delim(I(x), delim = "\t", show_col_types = FALSE, col_names = FALSE) |>
    dplyr::pull()
}

str(tc("3s03"))
#>  num 3000
str(parse_guess("3s03"))
#>  num 3000
str(vr("3s03"))
#>  num 3000

str(tc("0xa3"))
#>  chr "0xa3"
str(parse_guess("0xa3"))
#>  chr "0xa3"
str(vr("0xa3"))
#>  chr "0xa3"

str(tc("1,234.56"), digits.d = 7)
#>  num 1234.56
str(parse_guess("1,234.56"), digits.d = 7)
#>  num 1234.56
str(vr("1,234.56"), digits.d = 7)
#>  num 1234.56

# "I promise you this is a number with some other clutter lying around"
str(parse_number("3s03"))
#>  num 3
str(parse_number("0xa3"))
#>  num 0
str(parse_number("1,234.56"), digits.d = 7)
#>  num 1234.56
```

<sup>Created on 2026-03-18 with [reprex v2.1.1.9000](https://reprex.tidyverse.org)</sup>

**Comment by hadley** [comment]
@jennybc can we make the parser more strict here? It’s pretty surprising behaviour and while the conditions needed to trigger the bug are pretty rare, it would be good to close this loop hole.
